### ISSUE TITLE
rpc_init always enables CUDA if it's available

### ISSUE THREAD

**Issue by cimes-isi** [pytorch/pytorch#80141]
### 🐛 Describe the bug

When initializing RPC, it appears that CUDA is automatically initialized even if an empty device list (i.e., not `None`) is specified in the backend options.  While `None` implies default behavior, an empty list should conceptually mean that no CUDA devices are to be used by RPC, and therefore the RPC backend should not automatically init CUDA.  Setting the environment variable `CUDA_VISIBLE_DEVICES=""` avoids this particular problem, but then prevents the application from initializing/using CUDA for other operations not managed by RPC.

```python
import os
import torch
from torch.distributed import rpc
os.environ['MASTER_ADDR'] = 'localhost'
os.environ['MASTER_PORT'] = '29500'
opts = rpc.TensorPipeRpcBackendOptions(devices=[]) # empty list, not None
torch.cuda.is_initialized() # returns False
rpc.init_rpc('worker0', world_size=1, rank=0, rpc_backend_options=opts)
torch.cuda.is_initialized() # returns True
```

I'm running on a Jetson AGX Xavier device, with Ubuntu 18.04, Python 3.8.0 in a virtualenv, with a custom build of PyTorch v1.10.2.  The script below reports clang 6, but in fact PyTorch is compiled with clang 8.

Thanks!

### Versions

Collecting environment information...
PyTorch version: 1.10.0a0+git71f889c
Is debug build: False
CUDA used to build PyTorch: 10.2
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.6 LTS (aarch64)
GCC version: (Ubuntu/Linaro 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: 6.0.0-1ubuntu2 (tags/RELEASE_600/final)
CMake version: version 3.10.2
Libc version: glibc-2.27

Python version: 3.8.0 (default, Dec  9 2021, 17:53:27)  [GCC 8.4.0] (64-bit runtime)
Python platform: Linux-4.9.253-tegra-aarch64-with-glibc2.27
Is CUDA available: True
CUDA runtime version: Could not collect
GPU models and configuration: Could not collect
Nvidia driver version: Could not collect
cuDNN version: Probably one of the following:
/usr/lib/aarch64-linux-gnu/libcudnn.so.8.2.1
/usr/lib/aarch64-linux-gnu/libcudnn_adv_infer.so.8.2.1
/usr/lib/aarch64-linux-gnu/libcudnn_adv_train.so.8.2.1
/usr/lib/aarch64-linux-gnu/libcudnn_cnn_infer.so.8.2.1
/usr/lib/aarch64-linux-gnu/libcudnn_cnn_train.so.8.2.1
/usr/lib/aarch64-linux-gnu/libcudnn_ops_infer.so.8.2.1
/usr/lib/aarch64-linux-gnu/libcudnn_ops_train.so.8.2.1
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.22.4
[pip3] torch==1.10.0a0+git71f889c
[pip3] torchvision==0.12.0
[conda] Could not collect

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang @kwen2501 @jjlilley @mrzzd

**Comment by rohan-varma** [comment]
Verified that I can repro this.

@H-Huang can we fix it by applying the same logic of `None` to empty list?

**Comment by cimes-isi** [comment]
> Have a fix for this in #80180.
> 

Thanks for looking at this.

Looking at the current #80180, I'm not sure it's implemented in a way that has the intended effect, since I believe `None` and `[]` both evaluate to `False`.  The core of what I'm trying to say is that `None` is _not_ (and should not be treated) the same as an empty collection (`[]` in this case).  In my view:

- `devices=None` should result in some default behavior since `None` is the default argument value.  It's perhaps worth discussing what the default behavior _should_ be, but the default behavior in the current implementation appears to be: enable CUDA if it's available.
- `devices=[]` should result in doing nothing, i.e., CUDA should not be enabled even if it's available.

> @cimes-isi I am curious, is the automatic cuda init causing you issues? I thought the operation doesn't have a particular performance impact unless I am mistaken.

It has some secondary effects since I am currently checking if cuda is initialized to make decisions about moving models and data to the GPU.  I believe it also increases memory consumption, which is important on resource-constrained systems.  It might also cause problems for folks if their GPUs are detected but cannot be used, perhaps because other SW is using them, PyTorch doesn't correctly use them, or there is some kind of initialization failure which shouldn't actually matter if they don't want to use the GPU(s).

**Comment by H-Huang** [comment]
> ... I believe it also increases memory consumption, which is important on resource-constrained systems. ...

Good points. Thank you!

> It's perhaps worth discussing what the default behavior should be, but the default behavior in the current implementation appears to be: enable CUDA if it's available.

Yep, that is the current behavior. There isn't a particular reason that CUDA need to be initialized when no devices are passed (since devices are not updated after initialization for a statically sized RPC group), so the current behavior seems like a bug. The new default behavior will have CUDA not enabled by default or when no devices are passed in (`None` and `[]`).
